llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/SplitReduction.cpp

```diff
@@ -0,0 +1,319 @@
+  RankedTensorType outputIndicesExpandedType =
+      RankedTensorType::get(expandedResultShape, indicesElementType);
+
+  // Initialize the expanded output values
+  SmallVector<Value> dynSizes;
+  for (auto en : llvm::enumerate(valuesExpandedType.getShape())) {
+    if (en.value() == ShapedType::kDynamicSize) {
+      dynSizes.push_back(
+          rewriter.create<tensor::DimOp>(loc, valuesExpanded, en.index()));
+    }
+  }
+  Value initTensorOutputValues = rewriter.create<mlir::linalg::InitTensorOp>(
+      loc, dynSizes, outputValuesExpandedType.getShape(), valueElementType);
+  Value initTensorOutputIndices = rewriter.create<mlir::linalg::InitTensorOp>(
+      loc, dynSizes, outputIndicesExpandedType.getShape(), indicesElementType);
+
+  // Initialize indices to positive infinity and values to negative infinity
+  // for a top (maxk) comparison.
+  Attribute negInfAttr;
+  if (auto intType = valueElementType.dyn_cast<IntegerType>()) {
+    negInfAttr = rewriter.getIntegerAttr(
+        intType, APInt::getSignedMinValue(intType.getWidth()));
+  } else {
+    auto negApFloat =
+        APFloat::getInf(valueElementType.cast<FloatType>().getFloatSemantics(),
+                        /*Negative=*/true);
+    negInfAttr = rewriter.getFloatAttr(valueElementType, negApFloat);
+  }
+  Value negInf = rewriter.create<arith::ConstantOp>(loc, negInfAttr);
+  Attribute posInfAttr =
+      rewriter.getIntegerAttr(indicesElementType, APInt::getSignedMaxValue(32));
+  Value posInf = rewriter.create<arith::ConstantOp>(loc, posInfAttr);
+  Value negInfTensor =
+      rewriter.create<linalg::FillOp>(loc, negInf, initTensorOutputValues)
+          .result();
+  Value posInfTensor =
+      rewriter.create<linalg::FillOp>(loc, posInf, initTensorOutputIndices)
+          .result();
+
+  SmallVector<Type> parallelTopkResultTypes{outputValuesExpandedType,
+                                            outputIndicesExpandedType};
+  SmallVector<Value> parallelTopkIns{valuesExpanded};
+  SmallVector<Value> parallelTopkOuts{negInfTensor, posInfTensor};
+
+  // Parallel topk
+  auto parallelTopkOp = rewriter.create<iree_compiler::IREE::LinalgExt::TopkOp>(
+      loc,
+      /* resultTypes= */
+      parallelTopkResultTypes,
+      /* ins= */ parallelTopkIns,
+      /* outs= */ parallelTopkOuts, kDimParallel);
+  rewriter.cloneRegionBefore(topkOp.region(), parallelTopkOp.region(),
+                             parallelTopkOp.region().end());
+
+  return parallelTopkOp;
+}
+
+// Update the output indices from the parallel TopK with the correct offsets.
+// Each parallel computation uses implicit indices (starting from 0) during
+// selection, but the values are part of the large input space split into M =
+// splitReductionFn() ways. The following linalg.generic adds the appropriate
+// offset to reflect to values original position. "Updated pos" = "initial
+// pos" + m * "parallel computation index"
+Value offsetParallelIndices(Location loc, PatternRewriter &rewriter,
```

Comment:
I'd like to discuss this further. This is more evidence to show that the overall approach here has gaps. This is purely an artifact of the compiler and better spec/transformation of the operation would have made this step unnecessary.
I dont think I am going to block this per se, but pointing this out as an indicator of something not being right.